HW1/P2/P2.py

```diff
@@ -12,20 +20,36 @@ def mandelbrot(x, y):
     return iteration
 
 def sum_values_for_partitions(rdd):
-    'Returns (as an RDD) the sum of V for each partition of a (K, V) RDD'
+ #   'Returns (as an RDD) the sum of V for each partition of a (K, V) RDD'
     # note that the function passed to mapPartitions should return a sequence,
     # not a value.
     return rdd.mapPartitions(lambda part: [sum(V for K, V in part)])
 
 def draw_image(rdd):
-    '''Given a (K, V) RDD with K = (I, J) and V = count,
-    display an image of count at each I, J'''
-
+#    '''Given a (K, V) RDD with K = (I, J) and V = count,
+#    display an image of count at each I, J'''
     data = rdd.collect()
+    #
     I = np.array([d[0][0] for d in data])
     J = np.array([d[0][1] for d in data])
     C = np.array([d[1] for d in data])
+    #
     im = np.zeros((I.max() + 1, J.max() + 1))
     im[I, J] = np.log(C + 1)  # log intensity makes it easier to see levels
     plt.imshow(im, cmap=cm.gray)
     plt.show()
+    plt.savefig("Mandelbrot", dpi=2000, facecolor='w', edgecolor='w')
+
+q = range(0,4000)
+X = sc.parallelize(q,100)
+coordinates = X.flatMap(lambda x: [(x,y) for y in q]) # successfully creates pixel coordinates...
+
+data = coordinates.map(lambda x: (x,mandelbrot((x[1]/1000.-2),(x[0]/1000.-2))))
+draw_image(data)
+
+#draw_image(data) # this does indeed draw the mandelbrot set. takes a while.
+
+# sum_values_for_partitions doesn't work: Exception: It appears that you are attempting to broadcast an RDD or reference an RDD from an action or transformation. RDD transformations and actions can only be invoked by the driver, not inside of other transformations; for example, rdd1.map(lambda x: rdd2.values.count() * x) is invalid because the values transformation and count action cannot be performed inside of the rdd1.map transformation. For more information, see SPARK-5063.
+# can't seem to make it work.
+
+#test test test
```

Comment:
I remember having to play around with some of the given functions a bit to figure out how they were working, but sum_values_for_partitions worked for me with this code:

plt.hist(sum_values_for_partitions(result).collect())
plt.savefig('P2a_hist2.png')

where result should be equivalent to your variable data.
Hope it helps!